src/internal/hooks/use-performance-marks/index.ts

```diff
@@ -1,20 +1,44 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
 
-import { useEffect } from 'react';
-import { useUniqueId } from './use-unique-id';
-import { useEffectOnUpdate } from './use-effect-on-update';
-import { useModalContext } from '../context/modal-context';
+import { useEffect, useRef } from 'react';
+import { useIdFallback } from '../use-unique-id';
+import { useEffectOnUpdate } from '../use-effect-on-update';
+import { useModalContext } from '../../context/modal-context';
 
+/*
+This hook allows setting an HTML attribute after the first render, without rerendering the component.
+*/
+function useAttributeWithoutRerendering(elementRef: React.RefObject<HTMLElement>, value: string) {
+  const attributeName = 'data-analytics-performance-mark';
+
+  const attributeValueRef = useRef<string | undefined>();
+
+  useEffect(() => {
+    attributeValueRef.current = value;
+    elementRef.current?.setAttribute(attributeName, value);
+  }, [value, elementRef]);
+
+  return {
+    [attributeName]: attributeValueRef.current,
+  };
+}
+
+/**
+ * This function returns an object that needs to be spread onto the same
+ * element as the `elementRef`, so that the data attribute is applied
+ * correctly.
+ */
 export function usePerformanceMarks(
   name: string,
   enabled: boolean,
   elementRef: React.RefObject<HTMLElement>,
   getDetails: () => Record<string, string | boolean | number | undefined>,
   dependencies: React.DependencyList
 ) {
-  const id = useUniqueId();
+  const id = useIdFallback();
```

Comment:
Why this change is needed?

Reply:
Apart from the why, looking at the `useIdFallback()` hook it generates a random id per usage using its `counter`, does this mean for every component that gets rendered we'll have a different data analytics attribute?

How does it generally work when we try to collect performance about all of our buttons or tables for example and each one of those has a different analytics id?

Reply:
`useUniqueId` uses React's `useId` hook when available, which creates an ID that is unique only for that React tree. If there are multiple React trees on a page (e.g. a micro-frontend architecture), these IDs can collide. With the fallback hook, we generate a random ID, which is very unlikely to collide.

Yes, each component instance gets its own ID here, i.e. if there are two buttons on the page they will have different attribute values. The ID is generally intended to be used only in the direction from a performance mark to a DOM node. So if you want to find the performance of all Buttons and Tables, you would first query all performance marks from the browser, then for each mark you lookup the DOM node.